### PR TITLE
Stop benchmark when running contracts

### PR DESCRIPTION
**Pull request by kvinwang**

### CODE REVIEW

File: crates/phactory/src/system/gk.rs [hunk 1]
```diff
@@ -911,7 +911,25 @@ where
...
+             => {
+                let n_contracts = if let MiningReportEvent::HeartbeatV2 {
+                    n_contracts,
+                    ..
+                } = event {
+                    n_contracts
+                } else {
+                    0
+                };
+                // TODO.kevin: Deal with workers running contracts.
```
Comment (kvinwang) [inline crates/phactory/src/system/gk.rs:931]: @h4x3rotab How do we deal with workers running contracts?

Comment (h4x3rotab) [reply crates/phactory/src/system/gk.rs:931]: Haven't thought deeply. Generally we want the miner to keep get the same reward as mining normally. We use the iteration count to calculate `p_instant`, which affects the actual reward. When the miner is hosting the contract, we can turn its mining state from `MiningIdle` to `MiningActive`, and just give it the reward as 100% `p_init`. When the miner returns back to `MiningIdle`, we can refresh the iterator and make it back to normal.

Comment (h4x3rotab) [reply crates/phactory/src/system/gk.rs:931]: Btw, I saw `n_contracts` and `n_clusters`. I think we should treat `n_clusters > 0` as "active"

File: crates/phactory/src/benchmark.rs [hunk 2]
```diff
@@ -9,31 +10,38 @@ const UNIT: usize = 1;
...
-
-// If pruntime is synchronizing the history blocks, we consider it as not ready to benchmark.
-static READY_TO_BENCH: AtomicBool = AtomicBool::new(true);
+static FLAGS: Mutex<Flags> = Mutex::new(Flags::MINING_IDLE);
+
+bitflags::bitflags! {
+    // Any flag is set when the benchmark is paused.
+    #[derive(Serialize, Deserialize)]
+    pub struct Flags: u32 {
+        const MINING_IDLE = 1 << 0;
```
[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/phactory/src/benchmark.rs]: Renamed to BENCH_PAUSED.

Moreover, we removed the `MiningActive` state on the chain. Not sure whether we should rename the `MiningIdle` to `Mining`. Renaming it should be a breaking change to the front-end.

File: crates/phactory/src/system/gk.rs (hunk 1 above)
Comment (kvinwang) [reply crates/phactory/src/system/gk.rs]: Treating n_clusters > 0 as "active" and Using p_init to calculate the reword. Done

File: pallets/phala/src/mining.rs [hunk 3]
```diff
@@ -625,6 +630,7 @@ pub mod pallet {
 							Some(miner) => miner,
 							None => continue, // Skip non-existing miners
 						};
+						// The MiningActive is removed, is this logic need to be changed?
 						// Skip non-mining miners
 						if !miner_info.state.is_mining() {
```
Comment (kvinwang) [inline pallets/phala/src/mining.rs:635]: In the previous, `is_mining()` meant MiningIdle or MiningUnresponsive (MiningActive was excluded). Now it means MiningIdle (including previous MiningActive), MiningUnresponsive.

Comment (h4x3rotab) [reply pallets/phala/src/mining.rs:635]: No problem. We never have `Active` state in the past

**Review by h4x3rotab — APPROVED** [review]
LGTM

**Pull request merged**